README.md
# Mariana ![version](https://img.shields.io/visual-studio-marketplace/v/mightbesimon.mariana-sublime?label=)

[![sponsor](https://img.shields.io/badge/-Sponsor_Us_-%23d73a49)](https://github.com/sponsors/mightbesimon)
![downloads](https://img.shields.io/vscode-marketplace/i/mightbesimon.mariana-sublime?color=white&label=&logo=visualstudiocode&logoColor=blue)
![publish](https://github.com/mightbesimon/vscode-mariana/actions/workflows/publish.yml/badge.svg)

> I already tried the other ones so you don't need to.

Most faithful port of Sublime Mariana theme!

Wow this took way too long 😅

![preview](thumbnails/mariana.png)

# Todo List

<details>

<summary>
	<strong>click</strong> 👇 to see todo items 👉 <em>if you'd like to contribute</em> 👈
</summary>

🚧 diff editor

🚧 merge conflict

🚧 panel 50% done

🚧 menu bar

🚧 command center

🚧 notification 50% done

🚧 banner

🚧 extension 50% done

🚧 keybinding labels

🚧 keyboard shortcut table

🚧 debug colours

🚧 testing colours

🚧 welcome page

🚧 breadcrums

🚧 snippets

🚧 symbol icons

🚧 debug icons

🚧 charts

🚧 ports

🚧 extension colours

</details>

> more ideas welcome 🙂
>
> PRs welcome 👨‍🍳👌💋

# Acknowledgements

- **Mariana Theme** - Dmitri Voronianski, Sublime HQ Pty Ltd

# Contributors

- **Simon** - [mightbesimon](https://github.com/mightbesimon)
- [guillemap](https://github.com/guillemap) ✨
- [gijocode](https://github.com/gijocode)
- [Johnnie Gray](https://github.com/jcmgray)
- 👉 you? 👈

# Colour Reference

```python
class Mariana(metaclass=enum[Colour]):
	'''	mariana theme from sublime
	'''
	MARIANA  = hsla(210, 15, 22)
	DARK_0   = MARIANA.clone(l=11)
	DARK_1   = MARIANA.clone(l=13)
	DARK_2   = MARIANA.clone(l=19)
	MEDIUM_0 = MARIANA.clone()
	MEDIUM_1 = MARIANA.clone(l=40, a=75)
	MEDIUM_2 = MARIANA.clone(l=45)
	LIGHT_0  = hsla(220, 12, 69)
	LIGHT_1  = hsla(220, 28, 88)
	RED      = hsla(357, 79, 65)
	CORAL    = hsla( 13, 93, 66)
	ORANGE   = hsla( 32, 93, 66)
	YELLOW   = hsla( 40, 94, 68)
	MINT     = hsla(114, 31, 68)
	TEAL     = hsla(180, 36, 54)
	BLUE     = hsla(210, 50, 60)
	PURPLE   = hsla(300, 30, 68)
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

> Author: [mightbesimon](https://github.com/mightbesimon)

## v1.2.0 - 3 September 2023

- ✨ [feature]: add jupyter notebook colours
- 🧑‍💻 [dev]: use `venv` in export and install tasks
- 🧑‍💻 [dev]: set pre-launch task to export

## v1.1.2 - 31 December 2022

- ✨ [feature]: add bright black and white to terminal

## v1.1.1 - 12 December 2022

- 📚 [doc]: update badges and changelogs

## v1.1.0 - 24 November 2022

- 🚀 [CI/CD]: generate theme json file at build
- ✨ [feature]: terminal cursor to orange
- ✨ [feature]: orange button text to dark

## v1.0.0 - 31 October 2022

- ✨ [feature]: editor rulers
- ✨ [feature]: terminal logs

## v0.99.99 — 27 July 2022

Initial pre-release

- ✨ [feature]: marinana theme ported from sublime


colour/test.py
'''	Copyright © 2022 mightbesimon.com
	All rights reserved.

	Material belonging to others may have been
	used under Creative Commons Licence or with
	explicit or implicit permission.
'''

from decimal import Decimal
from unittest import TestCase, main
from . import Colour, hsla, rgba


################################################################
#######                   base classes                   #######
################################################################
class AssertColour:

	def assert_hsla(self,
		colour:hsla,
		h:int,
		s:Decimal,
		l:Decimal,
		a:Decimal
	) -> None:
		self.assertEqual(colour.h, h)
		self.assertEqual(colour.s, s)
		self.assertEqual(colour.l, l)
		self.assertEqual(colour.a, a)

	def assert_rgba(self,
		colour:rgba,
		r:int,
		g:Decimal,
		b:Decimal,
		a:Decimal
	) -> None:
		self.assertEqual(colour.r, r)
		self.assertEqual(colour.g, g)
		self.assertEqual(colour.b, b)
		self.assertEqual(colour.a, a)


################################################################
#######                      tests                       #######
################################################################
class TestHSLA(AssertColour, TestCase):

	def test_init(self) -> None:
		c1 = hsla(10, 10, 10)
		self.assertIsInstance(c1, Colour)
		self.assertIsInstance(c1, hsla)
		self.assert_hsla(c1, 10, 0.1, 0.1, None)

		c2 = hsla(s=1, a=50)
		self.assert_hsla(c2, None, 1, None, 0.5)

	def test_call(self) -> None:
		c1 = hsla(10, 10, 10)
		c2 = c1()
		c3 = c1(h=30)
		c4 = c3(a=0.5)

		self.assertFalse(c1 is c2)
		self.assert_hsla(c2, 10, 0.1, 0.1, None)
		self.assert_hsla(c3, 30, 0.1, 0.1, None)
		self.assert_hsla(c4, 30, 0.1, 0.1, 0.5)

	def test_clone(self) -> None:
		c1 = hsla(10, 10, 10)
		c2 = c1.clone()
		c3 = c1.clone(h=30)
		c4 = c3.clone(a=0.5)

		self.assertFalse(c1 is c2)
		self.assert_hsla(c2, 10, 0.1, 0.1, None)
		self.assert_hsla(c3, 30, 0.1, 0.1, None)
		self.assert_hsla(c4, 30, 0.1, 0.1, 0.5)

	def test_to_hsla(self) -> None:
		...

	def test_to_rgba(self) -> None:
		mint = hsla(114, 31, 68)
		self.assert_rgba(mint.to_rgba(), 153, 199, 148, None)

		orange = hsla(32, 85, 55)
		self.assert_rgba(orange.to_rgba(), 238, 147, 43, None)

	def test_to_hex(self) -> None:
		mint = hsla(114, 31, 68)
		self.assertEqual(mint.to_hex(), '#99c794')

		orange = hsla(32, 85, 55)
		self.assertEqual(orange.to_hex(), '#ee932b')


class TestRGBA(AssertColour, TestCase):

	def test_init(self) -> None:
		c = rgba(10, 10, 10)
		self.assertIsInstance(c, Colour)
		self.assertIsInstance(c, rgba)
		self.assert_rgba(c, 10, 10, 10, None)

		c = rgba(g=50, a=0.5)
		self.assert_rgba(c, None, 50, None, 0.5)


	def test_call(self) -> None:
		c1 = rgba(10, 10, 10)
		c2 = c1.clone()
		c3 = c1.clone(r=30)
		c4 = c3.clone(a=0.5)

		self.assertFalse(c1 is c2)
		self.assert_rgba(c2, 10, 10, 10, None)
		self.assert_rgba(c3, 30, 10, 10, None)
		self.assert_rgba(c4, 30, 10, 10, 0.5)

	def test_clone(self) -> None:
		c1 = rgba(10, 10, 10)
		c2 = c1.clone()
		c3 = c1.clone(r=30)
		c4 = c3.clone(a=0.5)

		self.assertFalse(c1 is c2)
		self.assert_rgba(c2, 10, 10, 10, None)
		self.assert_rgba(c3, 30, 10, 10, None)
		self.assert_rgba(c4, 30, 10, 10, 0.5)

	def test_to_hsla(self) -> None:
		mint = rgba(153, 199, 148)
		self.assert_hsla(mint.to_hsla(), 114, 0.31, 0.68, None)

		orange = rgba(238, 147, 43)
		self.assert_hsla(orange.to_hsla(), 32, 0.85, 0.55, None)

	def test_to_rgba(self) -> None:
		...

	def test_to_hex(self) -> None:
		mint = rgba(153, 199, 148, 100)
		self.assertEqual(mint.to_hex(), '#99c794')

		orange = rgba(238, 147, 43, 100)
		self.assertEqual(orange.to_hex(), '#ee932b')

		transparent = rgba(0, 0, 0, 0)
		self.assertEqual(transparent.to_hex(), '#00000000')


################################################################
#######                 MAIN STARTS HERE                 #######
################################################################
if __name__ == '__main__':
	main()


mariana.py
'''	Copyright © 2022 mightbesimon.com
	All rights reserved.

	Material belonging to others may have been
	used under Creative Commons Licence or with
	explicit or implicit permission.
'''

from lifehacks.colour import Colour
from lifehacks.colour.palette import Mariana
from lifehacks.metaclasses import enum


class ThemeReference:

	def __init__(self, filename:str) -> None:
		self.filename = filename

	def use_themes(self, *themes:enum[Colour]) -> 'ThemeReference':
		self.themes = themes
		return self

	def export_color_theme(self, filename:str) -> 'ThemeReference':
		with open(self.filename, 'r') as file:
			content = file.read()

		content = ( content
			.replace(': ', ':')
			.replace('\t', '' )
			.replace('\n', '' )
		)

		for theme in self.themes:
			for name, colour in theme:
				content = content.replace(name, colour.to_hex())

		with open(filename, 'w') as file:
			file.write(content)

		return self


################################################################
#######                 MAIN STARTS HERE                 #######
################################################################
if __name__ == '__main__':\
(
	ThemeReference(filename='themes/mariana-reference.json')
		.use_themes(Mariana)
		.export_color_theme(filename='themes/mariana-color-theme.json')
)


colour/colour.py
'''	Copyright © 2022 mightbesimon.com
	All rights reserved.

	Material belonging to others may have been
	used under Creative Commons Licence or with
	explicit or implicit permission.
'''

from abc import ABC, abstractmethod
from dataclasses import dataclass
from decimal import Decimal
from typing import Tuple


################################################################
#######                   base classes                   #######
################################################################
class Colour(ABC):
	'''	abstract base `Colour` class
	'''
	@abstractmethod
	def to_hsla(self) -> 'hsla':
		raise NotImplemented

	@abstractmethod
	def to_rgba(self) -> 'rgba':
		raise NotImplemented

	def to_hex(self) -> str:
		c = self.to_rgba()
		alpha = f'{round(c.a*255):02x}' if c.a is not None and c.a!=1 else ''
		return f'#{c.r:02x}{c.g:02x}{c.b:02x}{alpha}'

################################################################
#######                   dataclasses                    #######
################################################################
@dataclass
class hsla(Colour):
	'''	`Colour` object in hsla space

		optional `h`: hue `[0, 359]`
		optional `s`: saturation `[0, 100]` or `[0.0, 1.0]`
		optional `l`: lightness `[0, 100]` or `[0.0, 1.0]`
		optional `a`: alpha `[0, 100]` or `[0.0, 1.0]`
	'''
	h: int     = None	# hue        [0  , 359]
	s: Decimal = None	# saturation [0.0, 1.0]
	l: Decimal = None	# lightness  [0.0, 1.0]
	a: Decimal = None	# alpha      [0.0, 1.0]

	def __init__(self,
		h:int=None,
		s:Decimal=None,
		l:Decimal=None,
		a:Decimal=None,
	) -> None:
		'''	optional `h`: hue `[0, 359]`
			optional `s`: saturation `[0, 100]` or `[0.0, 1.0]`
			optional `l`: lightness `[0, 100]` or `[0.0, 1.0]`
			optional `a`: alpha `[0, 100]` or `[0.0, 1.0]`
		'''
		if h and not 0<=h<=359: raise ValueError(f'hue {h} not in range [0, 359]')
		if s and not 0<=s<=100: raise ValueError(f'saturation {s} not in range [0, 100] or [0.0, 1.0]')
		if l and not 0<=l<=100: raise ValueError(f'lightness {l} not in range [0, 100] or [0.0, 1.0]')
		if a and not 0<=a<=100: raise ValueError(f'alpha {a} not in range [0, 100] or [0.0, 1.0]')

		if h is not None: self.h = h
		if s is not None: self.s = s/100 if s>1 else s
		if l is not None: self.l = l/100 if l>1 else l
		if a is not None: self.a = a/100 if a>1 else a

	def __call__(self, **kwargs) -> 'hsla':
		return self.clone(**kwargs)

	def clone(self,
		h:int=None,
		s:Decimal=None,
		l:Decimal=None,
		a:Decimal=None,
	) -> 'hsla':
		'''	create a new instance of `hsla`,
			optionally modify value fields
		'''
		return self.__class__(
			h=h if h is not None else self.h,
			s=s if s is not None else self.s,
			l=l if l is not None else self.l,
			a=a if a is not None else self.a,
		)

	def to_hsla(self) -> 'hsla':
		'''return self'''
		return self

	def to_rgba(self) -> 'rgba':
		'''	[formula](https://www.rapidtables.com/convert/color/hsl-to-rgb.html)
		'''
		C = (1 - abs(2*self.l - 1)) * self.s
		X = C * (1 - abs(self.h/60%2 - 1))
		m = self.l - C/2

		table = {
			(0, 60): (C, X, 0),
			(60, 120): (X, C, 0),
			(120, 180): (0, C, X),
			(180, 240): (0, X, C),
			(240, 300): (X, 0, C),
			(300, 360): (C, 0, X),
		}

		for key, value in table.items():
			if not key[0]<=self.h<key[1]: continue
			r_, g_, b_ = value

			return rgba(
				r=round((r_+m) * 255),
				g=round((g_+m) * 255),
				b=round((b_+m) * 255),
				a=self.a,
			)

@dataclass
class rgba(Colour):
	'''	`Colour` object in rgba space

		optional `r`: red `[0, 255]`
		optional `g`: green `[0, 255]`
		optional `b`: blue `[0, 255]`
		optional `a`: alpha `[0, 100]` or `[0.0, 1.0]`
	'''
	r: int     = None	# red   [0  , 255]
	g: int     = None	# green [0  , 255]
	b: int     = None	# blue  [0  , 255]
	a: Decimal = None	# alpha [0.0, 1.0]

	def __init__(self,
		r:int=None,
		g:int=None,
		b:int=None,
		a:Decimal=None,
	) -> None:
		'''	optional `r`: red `[0, 255]`
			optional `g`: green `[0, 255]`
			optional `b`: blue `[0, 255]`
			optional `a`: alpha `[0, 100]` or `[0.0, 1.0]`
		'''
		if r and not 0<=r<=255: raise ValueError(f'red {r} not in range [0, 255]')
		if g and not 0<=g<=255: raise ValueError(f'green {g} not in range [0, 255]')
		if b and not 0<=b<=255: raise ValueError(f'blue {b} not in range [0, 255]')
		if a and not 0<=a<=100: raise ValueError(f'alpha {a} not in range [0, 100] or [0.0, 1.0]')

		if r is not None: self.r = r
		if g is not None: self.g = g
		if b is not None: self.b = b
		if a is not None: self.a = a/100 if a>1 else a

	def clone(self,
		r:int=None,
		g:int=None,
		b:int=None,
		a:Decimal=None,
	) -> 'rgba':
		'''	create a new instance of `rgba`,
			optionally modify value fields
		'''
		return self.__class__(
			r=r if r is not None else self.r,
			g=g if g is not None else self.g,
			b=b if b is not None else self.b,
			a=a if a is not None else self.a,
		)

	def normalise(self) -> Tuple[float, float, float]:
		'''	normalise to `(r, g, b)` between `0.0` and `1.0`
		'''
		return (self.r/255, self.g/255, self.b/255)

	def to_hsla(self) -> 'hsla':
		'''	[formula](https://www.rapidtables.com/convert/color/rgb-to-hsl.html)
		'''
		r_, g_, b_ = self.normalise()
		C_max = max(r_, g_, b_)
		C_min = min(r_, g_, b_)
		delta = C_max - C_min

		H = round(
			0 if delta==0 else
			60 * ((g_-b_)/delta % 6) if C_max==r_ else
			60 * ((b_-r_)/delta + 2) if C_max==g_ else
			60 * ((r_-g_)/delta + 4) if C_max==b_ else
			None
		)
		L = round((C_max+C_min) / 2, 2)
		S = round(delta / (1 - abs(2*L - 1)), 2)

		return hsla(H, S, L, self.a)

	def to_rgba(self) -> 'rgba':
		'''return self'''
		return self


colour/palette.py
'''	Copyright © 2022 mightbesimon.com
	All rights reserved.

	Material belonging to others may have been
	used under Creative Commons Licence or with
	explicit or implicit permission.
'''

from itertools import chain
from typing import List, Tuple
from . import Colour, hsla, rgba

################################################################
#######                    metaclass                     #######
################################################################
class MetaPalette(type):
	''' metaclass for `Palette`\n
		can be used as a `type` for type hinting\n
		e.g.
		```python
		def use_theme(self, theme:MetaPalette) -> None:
			self.theme:MetaPalette = theme
		```
	'''
	def __iter__(self) -> List[Tuple[str, Colour]]:
		'''	return all enum items from this enum class
			as well as from base classes\n
			e.g.
			```python
			for name, colour in Palette:
				content = content.replace(name, colour.to_hex())
			```
		'''
		iter_self = ( (name, attr)
			for name, attr in self.__dict__.items()
			if isinstance(attr, Colour)
		)
		try:
			return chain(iter_self, ( (name, attr)
				for name, attr in self.__base__
				if isinstance(attr, Colour)
			))
		except TypeError:
			return iter_self

	def __contains__(self, obj:Colour) -> bool:
		'''	check if obj is in this enum class
			as well as in base classes\n
			e.g.
			```python
			rgba(0,0,0) in Palette  # True
			```
		'''
		in_self = obj in [ attr
			for name, attr in self.__dict__.items()
			if isinstance(attr, Colour)
		]
		try:
			return in_self or obj in self.__base__
		except TypeError:
			return in_self

################################################################
#######                   base palette                   #######
################################################################
class Palette(metaclass=MetaPalette):
	''' base palette enum\n
		extend this enum class like so
		```python
		class Mariana(Palette):
			...
		```
	'''
	BLACK  = rgba(  0,   0,   0)
	WHITE  = rgba(255, 255, 255)
	SHADOW = BLACK.clone(a=50)
	TRANSPARENT = BLACK.clone(a=0)

################################################################
#######                     palettes                     #######
################################################################
class Mariana(Palette):
	'''	mariana theme from sublime
	'''
	MARIANA  = hsla(210, 15, 22)
	DARK_0   = MARIANA.clone(l=11)
	DARK_1   = MARIANA.clone(l=13)
	DARK_2   = MARIANA.clone(l=19)
	MEDIUM_0 = MARIANA.clone()
	MEDIUM_1 = MARIANA.clone(l=40, a=75)
	MEDIUM_2 = MARIANA.clone(l=45)
	LIGHT_0  = hsla(220, 12, 69)
	LIGHT_1  = hsla(220, 28, 88)
	RED      = hsla(357, 79, 65)
	CORAL    = hsla( 13, 93, 66)
	ORANGE   = hsla( 32, 93, 66)
	YELLOW   = hsla( 40, 94, 68)
	MINT     = hsla(114, 31, 68)
	TEAL     = hsla(180, 36, 54)
	BLUE     = hsla(210, 50, 60)
	PURPLE   = hsla(300, 30, 68)

class HealthNow:
	...

class Apple:
	# RED
	# YELLOW
	# GREEN
	LUCK = rgba(135, 188, 93)
